src/appleseed/foundation/utility/zipper.cpp

```diff
@@ -0,0 +1,345 @@
+        filename[filename.size() - 1] = '\0';
+
+        return string(&filename[0]);
+    }
+
+    string get_filepath(unzFile& zip_file, const string& unzipped_dir)
+    {
+        string filename = read_filename(zip_file);
+        return (bf::path(unzipped_dir) / bf::path(filename)).string();
+    }
+
+    void extract_current_file(unzFile& zip_file, const string& unzipped_dir)
+    {
+        const string filepath = get_filepath(zip_file, unzipped_dir);
+
+        if (is_zip_entry_directory(filepath))
+        {
+            bf::create_directories(filepath);
+            return;
+        }
+        else open_current_file(zip_file);
+
+        fstream out(filepath.c_str(), ios_base::out | ios_base::binary);
+        if (out.fail())
+            throw ZipException(("Can't open file " + filepath).c_str());
+
+        const size_t BUFFER_SIZE = 4096;
+        char buffer[BUFFER_SIZE];
+
+        do
+        {
+            const int read = read_chunk(zip_file, (char*) &buffer, BUFFER_SIZE);
+            out.write((char*) &buffer, read);
+        }
+        while (!unzeof(zip_file));
+
+        out.close();
+        unzip_close_current_file(zip_file);
+    }
+
+    void unzip(const string& zip_filename, const string& unzipped_dir)
+    {
+        try
+        {
+            bf::create_directories(unzipped_dir);
+
+            unzFile zip_file = unzOpen(zip_filename.c_str());
+            if (zip_file == 0)
+                throw ZipException(("Can't open file " + zip_filename).c_str());
+
+            unzGoToFirstFile(zip_file);
+
+            int has_next = UNZ_OK;
+            while (has_next == UNZ_OK)
+            {
+                extract_current_file(zip_file, unzipped_dir);
+                has_next = unzGoToNextFile(zip_file);
+            }
+
+            unzClose(zip_file);
+        }
+        catch (exception e)
+        {
+            bf::remove_all(unzipped_dir);
+            throw e;
+        }
+    }
+
+    void zip_close_current_file(zipFile& zip_file)
+    {
+        const int err = zipCloseFileInZip(zip_file);
+
+        if (err != ZIP_OK)
+            ZipException("Error while closing file in zip", err);
+    }
+
+    void write_chunk(zipFile& zip_file, char* buffer, const int chunk_size)
+    {
+        const int err = zipWriteInFileInZip(zip_file, buffer, chunk_size);
+
+        if (err < 0)
+        {
+            throw ZipException("Error while writing to zip", err);
+        }
+    }
+
+    void open_new_file_in_zip(zipFile& zip_file, string filename_in_zip, zip_fileinfo zip_file_info)
+    {
+        int err = zipOpenNewFileInZip(zip_file, filename_in_zip.c_str(), &zip_file_info,
+                                      0, 0, 0, 0, 0,
+                                      Z_DEFLATED, Z_DEFAULT_COMPRESSION);
+
+        if (err != ZIP_OK)
+            throw ZipException(("error while opening " + filename_in_zip + " in zipfile").c_str());
+    }
+
+    zip_fileinfo set_file_timestamp(const string& filename)
```

Comment:
The name of this function is not really correct: it does not set the timestamp of a file, instead it returns a properly initialized `zip_fileinfo` structure. May I suggest `make_zip_fileinfo()` instead?